packages/plugin/src/index.js

```diff
@@ -1,17 +1,42 @@
-import React from 'react'
+import React, { useEffect, useState } from 'react'
 import { VIS_TYPE_PIVOT_TABLE } from '@dhis2/analytics'
+import { useDataEngine } from '@dhis2/app-runtime'
 
+import { apiFetchLegendSet } from './api/legendSets'
 import ChartPlugin from './ChartPlugin'
 import PivotPlugin from './PivotPlugin'
 
 const VisualizationPlugin = props => {
+    const engine = useDataEngine()
+    const [legendSet, setLegendSet] = useState(null)
+
+    useEffect(() => {
+        const fetchLegendSet = async engine => {
+            if (
+                props.visualization.legendSet &&
+                props.visualization.legendSet.id
+            ) {
+                const response = await apiFetchLegendSet(
+                    engine,
+                    props.visualization.legendSet.id
+                )
+
+                if (response && response.legendSet) {
+                    setLegendSet(response.legendSet)
+                }
+            }
+        }
+
+        fetchLegendSet(engine)
+    }, [props.visualization.legendSet]) // eslint-disable-line react-hooks/exhaustive-deps
+
     if (
         !props.visualization.type ||
         props.visualization.type === VIS_TYPE_PIVOT_TABLE
     ) {
-        return <PivotPlugin {...props} />
+        return <PivotPlugin {...props} legendSet={legendSet} />
     } else {
-        return <ChartPlugin {...props} />
+        return <ChartPlugin {...props} legendSet={legendSet} />
```

Comment:
This will render the table or chart without the legend first (while it's loading) and then re-render it with the legend once the legendSet loads, is that the intent here?  One option to avoid this would be - 

```js
const VisualizationPlugin = props => {
    const engine = useDataEngine()
    const [legendSet, setLegendSet] = useState(null)

    const hasLegendSet = props.visualization.legendSet && props.visualization.legendSet.id

    useEffect(() => {
        const fetchLegendSet = async engine => {
            if (hasLegendSet) {
                const response = await apiFetchLegendSet(
                    engine,
                    props.visualization.legendSet.id
                )

                if (response && response.legendSet) {
                    setLegendSet(response.legendSet)
                }
            }
        }

        fetchLegendSet(engine)
    }, [props.visualization.legendSet]) // eslint-disable-line react-hooks/exhaustive-deps

    if (hasLegendSet && !legendSet) {
        // Until one of the children is rendered and calls onLoadingComplete,
        //   the app will continue to render the loading spinner
        return null
    }

    if(
        !props.visualization.type ||
        props.visualization.type === VIS_TYPE_PIVOT_TABLE
    ) {	
        return <PivotPlugin {...props} legendSet={legendSet} />
    } else {	
        return <ChartPlugin {...props} legendSet={legendSet} />
    }
}

Reply:
Yes, good point.
If the legendSet option is set, we know it before the first render happens, so no point in re-render.